felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -1931,7 +1949,35 @@ func (m *bpfEndpointManager) onServiceUpdate(update *proto.ServiceUpdate) {
 	ips = append(ips, update.ExternalIps...)
 
 	key := serviceKey{name: update.Name, namespace: update.Namespace}
-	m.dirtyServices[key] = ips
+
+	ips4 := make([]ip.V4CIDR, 0, len(ips))
+	for _, i := range ips {
+		cidr, err := ip.CIDRFromString(i + "/32")
+		if err != nil {
+			log.WithFields(log.Fields{"service": key, "ip": i}).Warn("Not a valid CIDR.")
+		} else if cidrv4, ok := cidr.(ip.V4CIDR); !ok {
+			log.WithFields(log.Fields{"service": key, "ip": i}).Debug("Not a valid V4 CIDR.")
+		} else {
+			ips4 = append(ips4, cidrv4)
+		}
+	}
+
+	// Check which IPs have been removed (no-op if we haven't seen it yet)
+	for _, old := range m.services[key] {
+		exists := false
+		for _, svcIP := range ips4 {
```

Comment:
This is only dealing with frontend service IPs, right?  I expect there are only ~10 of those at most so the `O(n^2)` loop is OK.  For backends, it wouldn't be OK since there can be 100k backends (if every pod in the cluster has a prometheus endpoint, for example).